beacon/sync/src/main/java/tech/pegasys/teku/beacon/sync/historical/ReconstructHistoricalStatesService.java

```diff
@@ -13,18 +13,118 @@
 
 package tech.pegasys.teku.beacon.sync.historical;
 
+import static tech.pegasys.teku.infrastructure.logging.StatusLogger.STATUS_LOG;
+
+import java.io.IOException;
+import java.util.Optional;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
 import tech.pegasys.teku.infrastructure.async.SafeFuture;
+import tech.pegasys.teku.infrastructure.exceptions.InvalidConfigurationException;
+import tech.pegasys.teku.infrastructure.unsigned.UInt64;
 import tech.pegasys.teku.service.serviceutils.Service;
+import tech.pegasys.teku.spec.Spec;
+import tech.pegasys.teku.spec.config.SpecConfig;
+import tech.pegasys.teku.spec.datastructures.state.beaconstate.BeaconState;
+import tech.pegasys.teku.spec.datastructures.util.ChainDataLoader;
+import tech.pegasys.teku.storage.api.StorageUpdateChannel;
+import tech.pegasys.teku.storage.client.CombinedChainDataClient;
 
 public class ReconstructHistoricalStatesService extends Service {
+  private static final Logger LOG = LogManager.getLogger();
+
+  private final CombinedChainDataClient chainDataClient;
+  private final Spec spec;
+  private final Optional<String> genesisStateResource;
+  private final StorageUpdateChannel storageUpdateChannel;
+
+  public ReconstructHistoricalStatesService(
+      final StorageUpdateChannel storageUpdateChannel,
+      final CombinedChainDataClient chainDataClient,
+      final Spec spec,
+      final Optional<String> genesisStateResource) {
+    this.storageUpdateChannel = storageUpdateChannel;
+    this.chainDataClient = chainDataClient;
+    this.spec = spec;
+    this.genesisStateResource = genesisStateResource;
+  }
 
   @Override
   protected SafeFuture<?> doStart() {
-    return SafeFuture.COMPLETE;
+    if (genesisStateResource.isEmpty()) {
+      return SafeFuture.failedFuture(
+          new IllegalStateException("Genesis state resource not provided"));
+    }
+
+    final String resource = genesisStateResource.get();
+    final BeaconState genesisState;
+    try {
+      genesisState = ChainDataLoader.loadState(spec, resource);
+    } catch (IOException e) {
+      LOG.error("Failed to load initial state", e);
+      return SafeFuture.failedFuture(
+          new InvalidConfigurationException(
+              "Failed to load initial state from " + resource + ": " + e.getMessage()));
+    }
+
+    return chainDataClient
+        .getInitialAnchor()
+        .thenAccept(
+            checkpoint -> {
+              if (checkpoint.isEmpty()) {
+                return;
+              }
+              applyBlocks(genesisState, checkpoint.get().getEpochStartSlot(spec));
+            });
+  }
+
+  public void applyBlocks(final BeaconState genesisState, final UInt64 anchorSlot) {
+    Context context = new Context(genesisState, SpecConfig.GENESIS_SLOT.plus(1), anchorSlot);
+    applyNextBlock(context).finish(STATUS_LOG::reconstructHistoricalStatesServiceFailedStartup);
```

Comment:
This probably needs to be a different status log method since it has failed while reconstructing states rather than during startup so the message would be slightly different.